orderbook.js

```diff
@@ -0,0 +1,41 @@
+const orderBook = (existingBook, incomingOrder) => {
+  // Check if exisitingBook has any values
+  if (existingBook.length <= 0) return [incomingOrder]
+
+  // Loop through exisiting book using a for..of loop. Also extracting iteration indexes
+  for (const [index, currentOrder] of existingBook.entries()) {
+    // Put special criteria into own variable, to clean up code
+    // eslint-disable-next-line max-len
+    const misMatchCriteria = ((incomingOrder.type === 'sell' && (incomingOrder.price <= currentOrder.price)) || (incomingOrder.type === 'buy' && incomingOrder.price >= currentOrder.price))
+    const matchingPriceCriteria = currentOrder.price === incomingOrder.price
+
+    if (currentOrder.type !== incomingOrder.type && misMatchCriteria) {
+      if (currentOrder.quantity === incomingOrder.quantity) {
+        existingBook.splice(index, 1)
+
+        return existingBook
+      } else if (currentOrder.quantity > incomingOrder.quantity && matchingPriceCriteria) {
+        currentOrder.quantity -= incomingOrder.quantity
+        existingBook.push(...existingBook.splice(index, 1))
+
+        return existingBook
+      } else if (incomingOrder.quantity > currentOrder.quantity && matchingPriceCriteria) {
+        incomingOrder.quantity -= currentOrder.quantity
+        existingBook.splice(index, 1)
```

Comment:
just fyi to note. splice changes the indexes and length of the array. problem can be solved without recursive if you manually change the for loop 'index' value by subtracting it by 1 after each splice